### DISCUSSION TITLE
Binder Error could not find key "key" in struct

### DISCUSSION THREAD

**Discussion by liquidcarbon** [duckdb/duckdb#11080] [General]
### What happens?

Hello, I recently updated DuckDB to 0.9.2 and I'm getting hard error on queries that I'm 99% (but not 100%) sure worked before.

When I run `COALESCE(A.Value, '')` I was expecting an empty string both when `A.Value` is NULL as well as when `A.Value` is absent.  Now it errors out.

I'm not sure how to get around that, because the structure of the JSON files I'm querying is quite variable and the keys are indeed often missing.

Related: https://stackoverflow.com/questions/77188196/is-it-possible-to-suppress-binder-error-could-not-find-key-in-duckdb

### To Reproduce

Create a STRUCT with a missing field and try to ask for it.

### OS:

Linux

### DuckDB Version:

0.9.2

### DuckDB Client:

python

### Full Name:

Alex

### Affiliation:

self

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by liquidcarbon** [comment]
Dug in more and found the culprit.
Nothing seems to have changed on the DuckDB side.

We changed the `FROM` clause in these queries to scan a more narrow set of JSON files that have less diverse schemas, and so the query engine is unaware that these struct keys _may_ exist.

Still, like the StackOverflow poster, I would prefer the behavior described (returning NULL for actual NULLS and missing keys alike).  Is this possible to arrange, or to have an option in `READ_JSON` for enabling this?

A possible hack-around: in every partition, create a fake JSON with all possible keys. 🤔

**Comment by szarnyasg** [comment]
Hello @liquidcarbon, can you please create a reproducible example for this issue?

**Comment by liquidcarbon** [comment]
Something like this:

```
cat /tmp/a.json

{"a": {"key": "value"}}
{"b": {"yek": "eulav"}}
```

```sql
SELECT
    a,
    a.key,
    b,
    b.yek,
    --b.key,
FROM '/tmp/a.json'
```
![image](https://github.com/duckdb/duckdb/assets/47034358/7cf2c79c-1ecd-4293-825d-1fe61c520ad4)

This works as expected, even though there's no `b` in the first row and no `a` in the second row.

But if you ask for `b.key` you get 
```
Binder Error: Could not find key "key" in struct

Candidate Entries: "yek"
```

And I think it should (have an option to) work and return a column of NULLs for `b.key`.

**Comment by szarnyasg** [comment]
Thanks! I'll discuss it with the team.

**Comment by serpiente** [comment]
Hello, is there an update on this or a workaround?

**Comment by soerenwolfers** [comment]
Imo, the important things here are (1) consistency (2) availability of *some* solution. 


For example, (1) you shouldn't make struct return nulls on missing keys if lists don't return nulls on missing indices as well (or you might even argue if tables don't return null values on non existing column specifiers) (2) it would be perfectly fine to have a function like `struct_extract` that takes an additional `default` argument.

**Comment by vyadav12** [comment]
For other query engines like trino/athena, the utility function COALESCE(A.Value, '') works perfectly fine but not with duckdb when A.Value does not exist. As you know schemas evolve over the time, so this is must FEATURE for a query engine. Any solution planning to implement ? or workaround

**Reply by serpiente** [reply]
Maybe this works? https://stackoverflow.com/questions/77188196/is-it-possible-to-suppress-binder-error-could-not-find-key-in-duckdb

**Comment by rodrigogonegit** [comment]
Any update on this?

**Comment by lacioffi** [comment]
Here's a dumb workaround:
Cast the STRUCT into a STRING or JSON, and work from there.

With a STRING, you can at least check if the field exists. Or you could manually parse it and extract the value using regex, I think.
With a JSON, it won't give a binder error if the field doesn't exist - it will return an empty result instead.

![image](https://github.com/user-attachments/assets/cbad8a0d-3015-4f50-bec3-526ea8a65e87)

**Comment by rodrigogonegit** [comment]
Is there any update on this? The linked alternative by using a python UDF would require some fiddling around (possibly highly inefficient) to make nested objects work.

**Comment by cyprismatt** [comment]
This is something I'm running into as well. There is a third party data source of json that may or may not contain specific keys from the result of an `unnest` function. I can't use duckdb until I know something like `extract_entry(record, 'key', NULL)` where `NULL` is a default value when the key does not exist.

**Reply by liquidcarbon** [reply]
try: figure out every possible key, create a fake JSON with made up values, name it 000.json, place it above all other files in each partition

```
a@AK-Desktop-6600:~$ cat tmp.json 
{"a": {"key": "value"}}
{"b": {"key": "bla", "yek": "eulav"}}
a@AK-Desktop-6600:~$ cat tmp2.json
{"a": {"key": "value"}}
{"b": {"yek": "eulav"}}

```


```
>>> duckdb.sql("SELECT a, b.key FROM '/home/a/tmp.json'")  # tmp.json comes first and sets schema
┌───────────────────────┬─────────┐
│           a           │   key   │
│ struct("key" varchar) │ varchar │
├───────────────────────┼─────────┤
│ {'key': value}        │ NULL    │
│ NULL                  │ bla     │
└───────────────────────┴─────────┘

>>> duckdb.sql("SELECT a, b.key FROM '/home/a/tmp*.json'")
┌───────────────────────┬─────────┐
│           a           │   key   │
│ struct("key" varchar) │ varchar │
├───────────────────────┼─────────┤
│ {'key': value}        │ NULL    │
│ NULL                  │ bla     │
│ {'key': value}        │ NULL    │
│ NULL                  │ NULL    │
└───────────────────────┴─────────┘

>>> duckdb.sql("SELECT a, b.key FROM '/home/a/tmp2.json'")  # tmp2.json alone doesn't work
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
duckdb.duckdb.BinderException: Binder Error: Could not find key "key" in struct

Candidate Entries: "yek"
```
